src/Microsoft.AspNet.OData.Shared/Query/Expressions/FilterBinder.cs

```diff
@@ -405,6 +328,35 @@ public virtual Expression BindCollectionResourceCastNode(CollectionResourceCastN
             return OfType(source, clrType);
         }
 
+        /// <summary>
+        /// Binds a <see cref="CollectionConstantNode"/> to create a LINQ <see cref="Expression"/> that
+        /// represents the semantics of the <see cref="CollectionConstantNode"/>.
+        /// </summary>
+        /// <param name="node">The node to bind.</param>
+        /// <returns>The LINQ <see cref="Expression"/> created.</returns>
+        public virtual Expression BindCollectionConstantNode(CollectionConstantNode node)
+        {
+            Expression source = null;
+            if (node.ItemType.IsInt16() || node.ItemType.IsInt32() || node.ItemType.IsInt64())
+            {
+                source = Expression.Constant(node.Collection.Select(a => a.Value).Cast<int>());
+            }
+            else if (node.ItemType.IsBoolean())
+            {
+                source = Expression.Constant(node.Collection.Select(a => a.Value).Cast<bool>());
+            }
+            else if (node.ItemType.IsString())
+            {
+                source = Expression.Constant(node.Collection.Select(a => a.Value).Cast<string>());
+            }
+            else
+            {
+                throw Error.NotSupported(SRResources.InNodeCollectionNotSupported);
```

Comment:
>Error.NotSupported(SRResources.InNodeCollectionNotSupported) [](start = 22, length = 60)

Perhaps a more specific error that says that only collections of primitives are supported with the "In" operator?